Add initial C API for circuit construction

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Building off the infrastructure for the sparse observable added in
PR #13445 this commit adds a C FFI for building Quantum Circuits. Right
now there is a function to create an empty circuit with n qubits and m
clbits and then a function to add standard gates to a circuit and then
also get the op counts out of a circuit. This is a start of the
functionality for a C API around interacting with circuits, later PRs
will expand this so we can have a more fully featured C API in the
future.

### Details and comments

Part of #13276

~This PR is based on top of #13986 and will need to rebased after that merges. To see the contents of just this PR you can look at: https://github.com/mtreinish/qiskit-core/compare/no-py-token-constructor-path...mtreinish:qiskit-core:c-api-circuit-rebased~

TODO:

- [x] Add release notes

mtreinish committed:
Add initial C API for circuit construction

Building off the infrastructure for the sparse observable added in
PR #13445 this commit adds a C FFI for building Quantum Circuits. Right
now there is a function to create an empty circuit with n qubits and m
clbits and then a function to add standard gates to a circuit and then
also get the op counts out of a circuit. This is a start of the
functionality for a C API around interacting with circuits, later PRs
will expand this so we can have a more fully featured C API in the
future.

Part of #13276

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,264 @@
+/// @param num_qubits The number of qubits the circuit contains.
+/// @param num_clbits The number of clbits the circuit contains.
+///
+/// @return A pointer to the created Circuit
```
I'll leave some nits since it'll be the public docs 😛 
```suggestion
/// @return A pointer to the created circuit.
```

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,264 @@
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let circuit = unsafe { const_ptr_as_ref(circuit) };
+
+    circuit.num_qubits() as u32
```
This is correct, but makes me think of another question: if we assume num_qubits to be `u32` everywhere, shouldn't `CircuitData::num_qubits` not return a `u32`?

mtreinish replied:
Looking at the code I expect it returns a usize because the method is also defined for python (and that was likely the original intent as the only consumer). It was less typing and simpler to just return a usize because it's doing `self.qubits.len()` and returning that to python which is going to convert it to a python int regardless of whether its a u32 or a usize. We can change it outside of this PR, it's a simple fix there but outside the scope of this PR.

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,264 @@
+/// # Example
+///
+///     QkCircuit *qc = qk_circuit_new(100);
+///     uint32_t num_qubits = qk_circuit_num_clbits(qc);  // num_qubits==100
```
```suggestion
///     QkCircuit *qc = qk_circuit_new(100, 50);
///     uint32_t num_clbits = qk_circuit_num_clbits(qc);  // num_clbits==50
```

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,264 @@
+///
+/// # Safety
+///
+/// Behavior is undefined if ``circt`` is not either null or a valid pointer to a
```
```suggestion
/// Behavior is undefined if ``circuit`` is not either null or a valid pointer to a
```

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,264 @@
+}
+
+/// @ingroup QkCircuit
+/// Append a standard gate to the circuit
```
Nit no. 2
```suggestion
/// Append a standard gate to the circuit.
```

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,264 @@
+/// @param circuit A pointer to the circuit to add the gate to
+/// @param gate The StandardGate to add to the circuit
+/// @param qubits The pointer to the array of ``uint32_t`` qubit indices to add the gate on
+/// @param params The pointer to the array of ``double`` values to use for the gate parameters
```
I'm annoyed at myself for commenting this
```suggestion
/// @param circuit A pointer to the circuit to add the gate to.
/// @param gate The StandardGate to add to the circuit.
/// @param qubits The pointer to the array of ``uint32_t`` qubit indices to add the gate on.
/// @param params The pointer to the array of ``double`` values to use for the gate parameters.
```

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,264 @@
+///     QkCircuit *qc = qk_circuit_new(100);
+///     qk_circuit_append_standard_gate(qc, HGate, *[0], *[]);
+///
+///
+/// # Safety
```
Not actually sure the spacing here needs to be exact, feel free to ignore 
```suggestion
/// # Safety
///
```

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,264 @@
+/// # Safety
+/// The ``qubits`` and ``params`` types are expected to be a non-null pointer to an array of
+/// ``uint32_t`` and ``double`` respectively where the length is matching the expectations for
+/// the standard gate. If the array is insufficently long the behavior of this function is
```
We should have a way of documenting how many parameters and qubits are needed, or to figure this out using functions/lookups. Especially if people (or we ourselves) would like to write automated tests which requires iterating over all gates or so.

What do you think of either e.g. adding a function that takes the stdgate enum? I assume we can't really expose a static array like we do in Rust

mtreinish replied:
We probably could expose the static array somehow, but having functions to look this up is a good idea. That's much simpler, and under the covers it will just look it up from the static array.

mtreinish replied:
I added functions to look this up in: https://github.com/Qiskit/qiskit/pull/14006/commits/06665c05c3c8a14e9ac6b0fb20fd70ae31714477

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,264 @@
+    params: *const f64,
+) -> ExitCode {
+    let circuit = unsafe { mut_ptr_as_ref(circuit) };
+    unsafe {
```
I'll channel my inner Jake: could you add a safety comment here? 😛

mtreinish replied:
Added in: https://github.com/Qiskit/qiskit/pull/14006/commits/c992f7ad0ee41c04e06d28258329c44623bbe0cf

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,264 @@
+///     qk_circuit_append_standard_gate(qc, HGate, *[0], *[]);
+///     qk_circuit_count_ops(qc);
+///
+/// # Safety
```
```suggestion
/// Return a list of string names for instructions in a circuit and their counts.
///
/// @param circuit A pointer to the circuit to get the counts for.
///
/// # Example
///
///     QkCircuit *qc = qk_circuit_new(100);
///     qk_circuit_append_standard_gate(qc, HGate, *[0], *[]);
///     qk_circuit_count_ops(qc);
///
/// # Safety
///
```

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,264 @@
+///
+/// # Safety
+///
+/// Behavior is undefined if ``op_counts`` is not the object returned by ``qk_circuit_count_ops``
```
```suggestion
/// @param op_counts The returned op count list from ``qk_circuit_count_ops``.
///
/// # Safety
///
/// Behavior is undefined if ``op_counts`` is not the object returned by ``qk_circuit_count_ops``.
```

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,264 @@
+
+#[repr(C)]
+pub struct OpCount {
+    name: *mut c_char,
```
Why does the `name` and below the `data` pointers need to be `mut`? Could they also be `const`, since it seems we don't change them Rust-side? Not that it would matter once we expose them, but at least that would let the rust compiler verify they aren't changed here 😄

mtreinish replied:
`data` and the other array types need to be mut because of the free implementation works. To coerce the pointer back into a `Box` that we then free we it's dropped we need the pointer to be mutable as the APIs all work with `* mut`. But for the string types I can make them `const` pointers and did in https://github.com/Qiskit/qiskit/pull/14006/commits/50b54087b37b42d8c958b775621b18dd4c81a168

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,264 @@
+/// Behavior is undefined ``circuit`` is not a valid, non-null pointer to a ``QkCircuit``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_circuit_count_ops(circuit: *const CircuitData) -> OpCounts {
```
This is a bit a different flow that we have compared to the `QkObsTerm/CSparseTerm` in the observable, where we ask the C user to allocate the struct and take it by reference to populate it. But after thinking about it I think this makes sense here since `OpCounts` needs to own the data, whereas `CSparseTerm` can just point to the data in the exisiting observable.

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,264 @@
+/// Behavior is undefined if ``op_counts`` is not the object returned by ``qk_circuit_count_ops``
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_circuit_free_count_ops(op_counts: OpCounts) {
```
Maybe 
```suggestion
pub unsafe extern "C" fn qk_opcounts_free(op_counts: OpCounts) {
```
since the scheme is `qk_<object>_free`?

mtreinish replied:
Renamed in: https://github.com/Qiskit/qiskit/pull/14006/commits/3e62ec6d8d4b68bb1521de0660659c214b7afa83

Cryoris commented on test/c/test_circuit.c:
```diff
@@ -0,0 +1,99 @@
+    qk_circuit_append_standard_gate(qc, QkStandardGate_XGate, qubits, params);
+    for (i = 0; i < 1000; i++) {
+        uint32_t qubits[1] = {i};
+        qk_circuit_append_standard_gate(qc, QkStandardGate_HGate, qubits, params);
```
``QkStandardGate_HGate`` is a lot of typing, but this what we get since we use the Rust enum, right? We definitely need to check out `cbindgen` via the Rust interface to see whether we can rename this to something shorter (same still holds for the observable bit terms). We can do this in a follow up/parallel work 🙂

mtreinish replied:
You fixed this so it's now `H` instead of `HGate`, and I renamed `QkStandardGate` to `QkGate` in https://github.com/Qiskit/qiskit/pull/14006/commits/66a989c819a7a86fc8d1362ec6fc422b9586fa53. So now it's `QkGate_H`.

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,264 @@
+///
+///
+/// # Safety
+/// The ``qubits`` and ``params`` types are expected to be a non-null pointer to an array of
```
In the tests you're passing ``params`` as ``NULL``, which is not allowed per documentation. Maybe we can just rephrase this stating that `params` and `qubits` can be NULL if the gate has no qubits (GlobalPhaseGate?) or no params?

mtreinish replied:
I updated the docs for this in: https://github.com/Qiskit/qiskit/pull/14006/commits/659c7196c236e84ce5c77d1fa7a1ab6685e42b00

Cryoris commented on test/c/test_circuit.c:
```diff
@@ -0,0 +1,99 @@
+        qk_circuit_append_standard_gate(qc, QkStandardGate_HGate, qubits, params);
+    }
+    QkOpCounts op_counts = qk_circuit_count_ops(qc);
+    if (op_counts.len != 3) {
```
op_counts and qc are not freed at failure 🙂

mtreinish replied:
The test doesn't fail :D

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/14006/commits/3e62ec6d8d4b68bb1521de0660659c214b7afa83 so it always cleans up after itself now.

Cryoris replied:
> The test doesn't fail :D

famous last words 😂 it's only been a few weeks writing C and I already have trust issues

Cryoris commented on test/c/test_circuit.c:
```diff
@@ -0,0 +1,99 @@
+        return EqualityError;
+    }
+    qk_circuit_free(qc);
+    return 0;
```
You could use `Ok` for consistency but it doesn't matter really 
```suggestion
    return Ok;
```

mtreinish replied:
I fixed this sort of in: https://github.com/Qiskit/qiskit/pull/14006/commits/3e62ec6d8d4b68bb1521de0660659c214b7afa83

Cryoris reviewed (commented):
This looks really cool! I left some comments below 🙂

mtreinish committed:
Add functions for standard instructions (except delay)

mtreinish committed:
Apply suggestions from code review

Co-authored-by: Julien Gacon <gaconju@gmail.com>

mtreinish committed:
Rename qk_circuit_free_count_ops() to qk_opcounts_free()

This also updates the test logic to make sure we always free even in
case of a failure.

mtreinish committed:
Add release note

mtreinish committed:
Add function to get instruction details

mtreinish committed:
Merge remote-tracking branch 'origin/main' into c-api-circuit-rebased

mtreinish committed:
Rename QkStandardGate -> QkGate

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,483 @@
+/// Behavior is undefined ``circuit`` is not a valid, non-null pointer to a ``QkCircuit``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_circuit_append_measure(
```
Should we just call this `qk_circuit_measure` (same for the other directives)? The `append` doesn't really add any information, or what do you think?

mtreinish replied:
I can drop append if you prefer. I named it this way because of the gate function I used append.

jakelishman replied:
Adding to the bikeshedding, but generally I'm with Julien - the `append` is just noise (and we don't have it in Python).  Similarly, I'd suggest `qk_circuit_append_standard_gate` just becomes `qk_circuit_gate`, since `StandardGate` should be _by far_ the most common thing to add. Once we support custom instructions, we can call that something like `qk_circuit_custom`.

mtreinish replied:
Renamed in: https://github.com/Qiskit/qiskit/pull/14006/commits/1240467ab1fba3a9ae3c57a514d9f272c235fe56

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,483 @@
+/// # Example
+///
+///     QkCircuit *qc = qk_circuit_new(100);
+///     qk_circuit_append_measure(qc, 0, 0);
```
The circuit doesn't have any classical bits, do we need to add one?
```suggestion
///     QkCircuit *qc = qk_circuit_new(100, 1);
///     qk_circuit_append_measure(qc, 0, 0);
```

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,483 @@
+/// @ingroup QkCircuit
+/// Append a measurement to the circuit
+///
+/// @param circuit A pointer to the circuit to add the gate to
```
```suggestion
/// @param circuit A pointer to the circuit to add the measurement to
```

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,483 @@
+    circuit: *mut CircuitData,
+    qubit: u32,
+) -> ExitCode {
+    let circuit = unsafe { mut_ptr_as_ref(circuit) };
```
I'm surprised cargo doesn't complain (maybe because of unsafe?) but the `circuit` should be declared `mut` if we push onto it, no? The same holds for the other methods 🙂

mtreinish replied:
Actually rust emits a warning saying the variable does not to be `mut` if you do `let mut circuit` here. It's likely because the return from `&mut` so it's explicitly mutable already.

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,483 @@
+///     qk_circuit_append_reset(qc, 0);
+///
+///
+/// # Safety
```
```suggestion
///     QkCircuit *qc = qk_circuit_new(100);
///     uint32_t qubits[5] = {0, 1, 2, 3, 4};
///     qk_circuit_append_barrier(qc, 5, qubits);
///
/// # Safety
/// 
```

Cryoris replied:
```suggestion
///     QkCircuit *qc = qk_circuit_new(100);
///     uint32_t qubits[5] = {0, 1, 2, 3, 4};
///     qk_circuit_append_barrier(qc, 5, qubits);
///
/// # Safety
/// 
```

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,483 @@
+///
+///     QkCircuit *qc = qk_circuit_new(100);
+///     qk_circuit_append_standard_gate(qc, HGate, *[0], *[]);
+///     qk_circuit_num_instructions(qc); // 1
```
```suggestion
/// @ingroup QkCircuit
/// Return the number of instructions in the circuit
///
/// @param circuit A pointer to the circuit to get the counts for.
///
/// # Example
///
///     QkCircuit *qc = qk_circuit_new(100);
///     qk_circuit_append_standard_gate(qc, QkGate_H, *[0], *[]);
///     qk_circuit_num_instructions(qc); // 1
```

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,483 @@
+///
+///     QkCircuit *qc = qk_circuit_new(100);
+///     qk_circuit_append_standard_gate(qc, HGate, *[0], *[]);
+///     qk_circuit_get_instruction(qc, 0); //
```
```suggestion
/// @ingroup QkCircuit
/// Return the number of instructions in the circuit
///
/// @param circuit A pointer to the circuit to get the counts for.
///
/// # Example
///
///     QkCircuit *qc = qk_circuit_new(100);
///     qk_circuit_append_standard_gate(qc, QkGate_H, *[0], *[]);
///     QkCircuitInstruction inst = qk_circuit_get_instruction(qc, 0);
```

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,483 @@
+/// Behavior is undefined if ``inst`` is not an object returned by ``qk_circuit_get_instruction``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_free_circuit_instruction(inst: CInstruction) {
```
This should be `qk_circuit_instruction_free` for consistency -- and should it take a pointer to a `CInstruction` otherwise the struct would get passed by value? Same question actually below for `OpCounts` 🙂

mtreinish replied:
It does take it by value, but that's because we're freeing the pointers that are contained in the struct not the struct itself. Each `CInstruction` has 4 pointers to a heap allocated arrays that we need to free. If we just drop the instruction those would be leaked. We can take a pointer to `CInstruction` but it functionally doesn't matter in this case because the struct itself isn't what we're freeing.

Cryoris replied:
But don't we also need to free the struct if it was allocated by `qk_circuit_get_instruction`? 🤔

mtreinish replied:
It's a directly returned stack allocated struct so we don't have to worry about freeing it, the memory will be reclaimed automatically like any local variable (unless the caller puts it on the heap but then it's their responsibility).

Cryoris replied:
Ah right because we don't Box::free anything we don't have to consume it again -- like the 1:1 mapping of `malloc` and `free` in C

mtreinish committed:
Apply suggestions from code review

Co-authored-by: Julien Gacon <gaconju@gmail.com>

mtreinish committed:
Add safety comments

mtreinish committed:
Add methods to lookup num qubits and num params for gates

mtreinish committed:
Update docs around null pointers for qubits and params on gate append function

mtreinish committed:
Fix clang-format-14 error

Locally I was using clang-format-19 while in CI we run with
clang-format-14. Version 14 was flagging this difference while version
19 was happy. This commit makes the change to make version 14 happy in
CI.

raynelfss commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,544 @@
+///
+/// # Safety
+///
+/// Behavior is undefined ``circuit`` is not a valid, non-null pointer to a ``QkCircuit``.
```
I realized this far into reading that this docstring is missing an "if"? There's also many instances of that so I'll just give you this one first 😄 
```suggestion
/// Behavior is undefined if ``circuit`` is not a valid, non-null pointer to a ``QkCircuit``.
```

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14006/commits/92a4a8f5d4cf08600b9d875534948f41fd693ce5

raynelfss commented on releasenotes/notes/c-api-circuit-3ecb43d396425bed.yaml:
```diff
@@ -0,0 +1,5 @@
+---
+features_circuits:
+  - |
+    The C API for Qiskit has been extended with support for building and
+    interacting with quantum circuits.
```
Maybe we should mention that the support of this API is limited ?
```suggestion
    The C API for Qiskit has been extended with limited support for building and
    interacting with quantum circuits.
```

mtreinish replied:
I'm not actually sure what limited would mean here. The API lets you build and interact with a quantum circuit. It's not the same as the python API but there will be continued incremental improvements to the API over time. We're also 3 months out from the final 2.1.0 release so it's hard to say how limited the API will be when we finish.

raynelfss commented on test/c/test_circuit.c:
```diff
@@ -0,0 +1,554 @@
+        }
+    }
+    return Ok;
+}
```
Would it be possible for us to iterate through the values of the Enum itself? AFAIK we could do something like:
```c
for (QkGate gate = X; i < RC3X; i++) {
    ...
}
```
But then again there's no perfect way of knowing the end value of an enum unless added explicitly. But in case we were to hypothetically add a new standard gate, we wouldn't necessarily have to update this test manually.

mtreinish replied:
I can change the integers to use the gate enum variants, but in practice I think trying to make this robust against future gates won't really be feasible. It would depend on semantics being applied to the enum ordering which isn't a guarantee, we have a convention now. But if we add gates that definitely won't hold because we're unlikely to ever change the existing order of gates in the future because it would require very large refactoring, more than just these tests.

raynelfss commented on test/c/test_circuit.c:
```diff
@@ -0,0 +1,554 @@
+    return Ok;
+}
+
+int test_get_gate_counts_bv_no_measure() {
```
Someone really likes Bernstein-Vazirani :)

raynelfss reviewed (commented):
I gave a quick review to what you've worked on so far. I feel like it makes sense to me even if the API is still very limited. I would suggest adding any additional support for other operations within the circuit in subsequent PRs.

Most of the observations here are just docstring comments :)

mtreinish committed:
Merge remote-tracking branch 'origin/main' into c-api-circuit-rebased

mtreinish committed:
Add missing if to docstrings

mtreinish committed:
Merge remote-tracking branch 'origin/main' into c-api-circuit-rebased

mtreinish committed:
Fix merge

mtreinish committed:
Fix macOS tests

mtreinish committed:
Fix cformat

Cryoris commented on test/c/CMakeLists.txt:
```diff
@@ -19,12 +19,13 @@ file (
 # test file.
 create_test_sourcelist (source_files test_driver.c ${discovered_tests})
 
+add_library(common common.c)
```
So I'm not sure if this fixes it on windows (locally it still works on my mac) but it seems like windows is not finding the library (again), maybe this fixes it?
```suggestion
add_library(common common.c)
install(TARGETS common RUNTIME DESTINATION ${CMAKE_BINARY_DIR})
```
which should install the dll into the path of the executable (see also [CMake docs](https://cmake.org/cmake/help/book/mastering-cmake/chapter/Install.html)). We could also consider putting this inside a `if (MSVC)` block, though it didn't seem to affect running tests for me.

mtreinish committed:
Update test/c/CMakeLists.txt

Co-authored-by: Julien Gacon <gaconju@gmail.com>

Cryoris commented on test/c/CMakeLists.txt:
```diff
@@ -19,12 +19,14 @@ file (
 create_test_sourcelist (source_files test_driver.c ${discovered_tests})
 
+add_library(common common.c)
+install(TARGETS common RUNTIME DESTINATION ${CMAKE_BINARY_DIR})
```
With a static lib we shouldn't need to worry about placing the DLLs in the right location on windows
```suggestion
add_library(common STATIC common.c)
```

Cryoris committed:
Try static lib for common

mtreinish committed:
Fix windows build warnings

mtreinish committed:
Merge remote-tracking branch 'origin/main' into c-api-circuit-rebased

mtreinish committed:
Merge remote-tracking branch 'origin/main' into c-api-circuit-rebased

mtreinish committed:
Try adding the parent path to the path on windows

mtreinish committed:
Use const pointers for strings in structs

raynelfss commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,544 @@
+/// @ingroup QkCircuit
+/// Return the number of instructions in the circuit
+///
+/// @param circuit A pointer to the circuit to get the counts for.
```
```suggestion
/// @param circuit A pointer to the circuit to get the number of operations of.
```

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/14006/commits/4f0e9660cb561bf765aa289b47b858eb2c1aea4a

raynelfss commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,544 @@
+/// Return the number of instructions in the circuit
+///
+/// @param circuit A pointer to the circuit to get the counts for.
+///
```
This docstring doesn't seem to be correct. If I understood correctly, this function access an instruction by index.

raynelfss replied:
Also:
```suggestion
/// @ingroup QkCircuit
/// Return the number of instructions in the circuit.
///
/// @param circuit A pointer to the circuit to be accessed.
```

mtreinish replied:
You're correct this docstring is incorrect it's a copy and paste of a different function and I forgot to update it.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/14006/commits/4f0e9660cb561bf765aa289b47b858eb2c1aea4a

raynelfss commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,544 @@
+}
+
+/// @ingroup QkCircuit
+/// Append a reset to the circuit
```
Might be better to specify that we're adding a reset gate?
```suggestion
/// Append a reset gate to the circuit
```

mtreinish replied:
A reset is not a gate since it's not a unitary operation.

raynelfss commented on test/c/common.c:
```diff
@@ -0,0 +1,44 @@
+    } else {
+        msg = "FAILED with unknown error";
+    }
+    fprintf(stderr, "--- %-30s: %s\n", name, msg);
```
Would it make sense to increase the padding to a little higher than 30 since some of the methods have longer names?

mtreinish replied:
It doesn't bother me too much that the Ok's are not aligned for the longer function names. Like: https://github.com/Qiskit/qiskit/actions/runs/14495258916/job/40661239698#step:4:370 looks fine to me

We can expand it, but I'd rather do that in a separate PR. I didn't really edit the code in this function I only split it out to support multiple test files.

raynelfss replied:
That's okay then, it doesn't bother me too much either, but I wanted to bring it up here :)

raynelfss commented on test/c/test_circuit.c:
```diff
@@ -0,0 +1,551 @@
+    size_t num_instructions = qk_circuit_num_instructions(qc);
+    qk_circuit_free(qc);
+
+    return (num_qubits != 0 || num_clbits != 0 || num_instructions != 0) ? EqualityError : Ok;
```
I think that separating these might be a bit better. Although I do understand we don't necessarily have tracebacks here. We could add a `print` statement in the future that says which of these might have failed.

mtreinish replied:
Updated in: https://github.com/Qiskit/qiskit/pull/14006/commits/296190d87fc99f097f33a057a4f1a02307654054

raynelfss commented on test/c/test_circuit.c:
```diff
@@ -0,0 +1,551 @@
+}
+
+int test_no_gate_1000_bits(void) {
+    QkCircuit *qc = qk_circuit_new(1000, 1000);
+    uint32_t num_qubits = qk_circuit_num_qubits(qc);
+    uint32_t num_clbits = qk_circuit_num_clbits(qc);
+    size_t num_instructions = qk_circuit_num_instructions(qc);
+    qk_circuit_free(qc);
```
See https://github.com/Qiskit/qiskit/pull/14006#discussion_r2047487020

mtreinish replied:
https://github.com/Qiskit/qiskit/pull/14006/commits/296190d87fc99f097f33a057a4f1a02307654054

raynelfss reviewed (commented):
I think this looks almost ready to merge. I have a couple docstring comments and a question about tests :)

mtreinish committed:
Improve docstrings for C API

mtreinish committed:
Improve equality check failures in sanity check tests

mtreinish committed:
Merge remote-tracking branch 'origin/main' into c-api-circuit-rebased

mtreinish committed:
Fix rustfmt

mtreinish committed:
Simplify cmake config for tests

jakelishman commented on crates/cext/src/circuit.rs:
```diff
@@ -0,0 +1,566 @@
+/// Behavior is undefined if ``inst`` is not an object returned by ``qk_circuit_get_instruction``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_free_circuit_instruction(inst: CInstruction) {
```
The name here should be `qk_circuit_instruction_free` for consistency, I think?

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/14006/commits/1240467ab1fba3a9ae3c57a514d9f272c235fe56

jakelishman reviewed (commented):
I didn't actually review the implementations, I just joined in with the bikeshedding.

mtreinish committed:
Apply naming suggestions

raynelfss commented on crates/cext/src/pointers.rs:
```diff
@@ -0,0 +1,48 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
```
```suggestion
// (C) Copyright IBM 2025
```

mtreinish replied:
The contents of this file were just moved from the original unchanged and they had a 2024 copyright header so I just preserved that.

raynelfss commented on test/c/common.c:
```diff
@@ -0,0 +1,44 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024.
```
```suggestion
// (C) Copyright IBM 2025.
```

raynelfss reviewed (approved):
LGTM! (except for the minor dates detail)

Cryoris commented on test/c/test_circuit.c:
```diff
@@ -22,27 +22,52 @@
 
-    return (num_qubits != 1000 || num_clbits != 1000 || num_instructions != 0) ? EqualityError : Ok;
+    if (num_qubits != 1000) {
+        printf("The number of qubits %d is not 1000", num_qubits);
```
It would be nice to standardize this in the future. I'm wondering if this is printed somehow in order with CTest's output or if it is shown at another point and the origin is hard to pin down.

mtreinish replied:
IIRC, I used print debugging for developing the tests in the first place and ctest put the output somewhere I could find. But if we need to make improvements here we can definitely do it in a follow up.

Cryoris commented on test/c/CMakeLists.txt:
```diff
@@ -32,12 +31,7 @@ target_link_libraries (test_driver ${qiskit} common)
-
-    # Get the Python runtime library, which has been stored into a config file during the 
-    # crate build process. This is to ensure we link against the same library that PyO3 is using.
-    file(READ "${CMAKE_SOURCE_DIR}/target/pyo3_python.config" pyo3_config)
```
Does the `pyo3_python.config` file still have value if we don't use it here? We could also remove it (fine to do as follow up) if we don't need it anymore

mtreinish replied:
Probably not, lets look to drop it in a follow-up

mtreinish replied:
I removed this in: https://github.com/Qiskit/qiskit/pull/14239

Cryoris reviewed (approved):
Two small comments (that can be skipped) -- LGTM!
